Mathlib/Topology/ContinuousFunction/CFC.lean

```diff
@@ -0,0 +1,1111 @@
+    [Module ℂ E] [SMul M E] [IsScalarTower ℂ M E] : IsScalarTower ℝ M E where
+  smul_assoc x _ _ := (smul_assoc (x : ℂ) _ _ : _)
+
+-- this is not so crazy, we already have the `•` in `Module.complexToReal`
+noncomputable instance Algebra.complexToReal {A : Type*} [Semiring A] [Algebra ℂ A] : Algebra ℝ A :=
+  RestrictScalars.algebra ℝ ℂ A
+
+/-- Post-composition with a continuous star algebra homomorphism is a star algebra homomorphism
+between spaces of continuous maps. -/
+@[simps]
+def ContinuousMap.compStarAlgHom (X : Type*) {R B C : Type*} [TopologicalSpace X] [CommSemiring R]
+    [Semiring B] [Algebra R B] [Star B] [TopologicalSpace B] [TopologicalSemiring B]
+    [ContinuousStar B] [Semiring C] [Algebra R C] [Star C] [TopologicalSpace C]
+    [TopologicalSemiring C] [ContinuousStar C] (φ : B →⋆ₐ[R] C) (hφ : Continuous φ) :
+    C(X, B) →⋆ₐ[R] C(X, C) where
+  toFun f := (⟨φ, hφ⟩ : C(B, C)).comp f
+  map_one' := ContinuousMap.ext fun _ => map_one φ
+  map_mul' f g := ContinuousMap.ext fun x => map_mul φ (f x) (g x)
+  map_zero' := ContinuousMap.ext fun _ => map_zero φ
+  map_add' f g := ContinuousMap.ext fun x => map_add φ (f x) (g x)
+  commutes' r := ContinuousMap.ext fun _x => AlgHomClass.commutes φ r
+  map_star' f := ContinuousMap.ext fun x => map_star φ (f x)
+
+open scoped NNReal
+
+instance : StarRing ℝ≥0 where
+  star := id
+  star_involutive _ := rfl
+  star_mul := mul_comm
+  star_add _ _ := rfl
+
+instance : TrivialStar ℝ≥0 where
+  star_trivial _ := rfl
+
+instance : ContinuousStar ℝ≥0 where
+  continuous_star := continuous_id
+
+instance : StarModule ℝ≥0 ℝ where
+  star_smul := by simp only [star_trivial, eq_self_iff_true, forall_const]
+
+theorem Polynomial.X_toContinuousMap (R : Type*) [Semiring R] [TopologicalSpace R]
+    [TopologicalSemiring R] : (X : R[X]).toContinuousMap = ContinuousMap.id R := by
+  ext
+  simp
+
+-- these aliases should exist but don't
+alias ⟨spectrum.not_isUnit_of_zero_mem, spectrum.zero_mem⟩ := spectrum.zero_mem_iff
+alias ⟨spectrum.isUnit_of_zero_not_mem, spectrum.zero_not_mem⟩ := spectrum.zero_not_mem_iff
+
+lemma spectrum.subset_singleton_zero_compl (R : Type*) {A : Type*} [CommSemiring R] [Ring A]
+    [Algebra R A] {a : A} (ha : IsUnit a) : spectrum R a ⊆ {0}ᶜ :=
+  Set.subset_compl_singleton_iff.mpr <| spectrum.zero_not_mem R ha
+
+/-- `x ↦ x⁻¹` bundled as a continuous map whose domain is a set not containing zero. -/
+@[simps]
+def ContinuousMap.inv₀On {G : Type*} [Zero G] [Inv G] [TopologicalSpace G] [HasContinuousInv₀ G]
+    {s : Set G} (hs : 0 ∉ s) : C(s, G) :=
+  ⟨(·⁻¹), (continuousOn_inv₀.mono <| Set.subset_compl_singleton_iff.mpr hs).restrict⟩
+
+end prereqs
+
+/-!
+## Definitions
+-/
+
+/-- This class exists because under modest hypotheses, we can get a `Subsingleton` instance for it.
+In general, one should create instances of `CFCClass` instead. -/
+@[ext (flat := false)]
+class CFCCoreClass (R : Type*) {A : Type*} [CommSemiring R] [StarRing R] [TopologicalSpace R]
```

Comment:
See [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/!4.233360.20new.20ext.20lemma.20generation/near/351104478). Basically, it means that the generated `ext` lemma takes as a hypothesis that the `StarAlgHom`s are equal, instead of that the underlying functions are equal. This way we can use `ext` lemmas specific to `StarAlgHom` afterwards.